### PR TITLE
Allowing getting sources from compile tasks

### PR DESCRIPTION
**Pull request by ting-yuan**
instead of default source set and its dependencies.

There are 3 ways (that I know of) to register extra sources to
compilation tasks:

  1. Add to default source sets.
  2. Add to dependencies of default source sets.
  3. Add to compile task directly, e.g., via AbstractCompile.source

KSP picked up from the first 2 only. Now it also does third.

### CODE REVIEW

**Comment by ting-yuan** [comment]
This is consistent with kaptGenerateStubsTask by the way.

File: gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt [hunk 1]
```diff
@@ -297,7 +297,13 @@ class KspGradleSubplugin @Inject internal constructor(private val registry: Tool
                 classOutputDir,
                 resourceOutputDir
             )
-            kotlinCompilation.allKotlinSourceSets.forEach { sourceSet -> kspTask.source(sourceSet.kotlin) }
+            kspTask.source(kotlinCompileTask.source)
```
Comment (gavra0) [inline gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt:300]: This may be tricky as kotlin compile task will have java/kotlin sources that are generated by ksp processors. In `kaptGenerateStubsTask` this is avoided by filtering out sources that are in the task output directories.

What is the motivation for this change, typically sourceSets should contain all information, and plugins should avoid adding sources directly to the task?

Comment (ting-yuan) [reply gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt:300]: I was thinking about minimizing surprises for users moving from kapt. This particular case / bug report is about sources added via `BaseVariant.registerJavaGeneratingTask` by other plugins.

Having said that, this behavior isn't quite right. We should probably not picking up generated sources at all.

**Review by neetopia — APPROVED** [review]

**Comment by ting-yuan** [comment]
An option is added to switch between the behaviors. Default is the old behavior.

File: gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt (hunk 1 above)
Comment (gavra0) [reply gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt]: With the toggle, this LGTM. We can also follow up with users that do enable it, to understand their use-cases a bit better.

**Review by gavra0 — APPROVED** [review]
